proposed/2022/vulnerabilities-in-restore.md

```diff
@@ -0,0 +1,391 @@
+
+#### Vulnerability warnings in  the solution explorer
+
+Given that restore would be raising a warning, the vulnerability information will automatically appear in the Solution Explorer.
+This is the primary motivation for having 1 error message per package.
+
+### Technical explanation
+
+During every real restore, i.e. one that regenerates the assets file, NuGet will check all sources that provide vulnerability information and raise warnings for the packages that do.
+
+*Which sources can provide vulnerability information?*
+Only V3 http sources can provide vulnerability information. No additional means of providing vulnerability information are proposed here.
+
+#### Technical background
+
+To best understand the technical proposal, it's important to understand the [NuGet package installation process](https://docs.microsoft.com/en-us/nuget/concepts/package-installation-process).
+
+When a package is seen, the client talks to sources and downloads whichever versions are needed. When talking to V3 sources, the client use the `package content` [resource](https://docs.microsoft.com/en-us/nuget/api/package-base-address-resource), which is a static resource that only has information about the available package versions. Only what is contained *within* the package itself affect restore. Nothing else. This allows an equivalence between `http` and `local` sources.
+
+#### Dedicated vulnerability information resource
+
+Checking vulnerability will always take a performance hit, but we want to minimize that. We can:
+
+- Make fewer http calls.
+- Minimize the amount of data we are handling.
+
+Both of these can be achieved by adding a means to understand whether a source supports vulnerability. We'd add a resource such as the following:
+
+```json
+    {
+      "@id": "https://nikolchevulntest.blob.core.windows.net/newcontainer/vulnerabilityindex.json",
+      "@type": "VulnerabilityInfo/6.3.0",
+      "comment": "The endpoint for discovering information about vulnerabilities of packages in this package source."
+    },
+```
+
+When NuGet restores, we check whether a source supports vulnerabilities and if so it can go through a protocol of getting this information.
+
+The client needs all the data to make decisions about vulnerability. In order to not have the client redownload a file and limit the total number of HTTP calls we will ever make, the vulnerability resource will be an index that can contain multiple pages.
+
+- The vulnernability resource **must** be an array of objects.
+- Each object **must** contain:
+  - `@name`, a user friendly name for the page, **must** be unique.
+  - `@id`, the url that contains the data.
+  - `@updated`, a UTC timestamp when the content at `@id` was updated last.
+  - `comment`, a user friendly description.

```

Comment:
What's the value for this field? Does this really need to be required? Will this be noisy repetitive values?